File: Hard/51.N-Queens/dfs.cpp
class Solution {
public:
    bool noAtk(vector<string> grid, int row, int col){
        // check col 
        for(int i = 0; i < grid.size(); i++)
            if(grid[i][col] == 'Q')
                return false;
        // check diagonal left
        for(int i = row - 1, j = col - 1; i >= 0 && j >= 0; i--, j--)
            if(grid[i][j] == 'Q')
                return false;
        // check diagonal right
        for(int i = row - 1, j = col + 1; i >= 0 && j < grid.size(); i--, j++)
            if(grid[i][j] == 'Q')
                return false;
        return true;
    }
    void helper(vector<vector<string>> &res, vector<string> grid, int row){
        if(row == grid.size()){
            res.push_back(grid);
            return;
        }
        for(int col = 0; col < grid.size(); col++){
            if(noAtk(grid, row, col)){
                grid[row][col] = 'Q';
                helper(res, grid, row + 1);
                grid[row][col] = '.';
            }
        }
    }
    vector<vector<string>> solveNQueens(int n) {
        
        vector<vector<string>> res;
        vector<string> grid(n, string(n, '.'));
        
        helper(res, grid, 0);
        
        return res;
        
    }
};

File: Go/Easy/463.IslandPerimeter/dfs.go
func islandPerimeter(grid [][]int) int {
    for i := 0; i < len(grid); i++{
        for j := 0; j < len(grid[i]); j++{
            if grid[i][j] == 1{
                return helper(grid, i, j)
            }
        }
    }
    return 0
}

func helper(grid [][]int, row int, col int) int {
    if !isValid(grid, row, col) || grid[row][col] == 0{
        return 1
    }
    if grid[row][col] == -1{
        return 0
    }
    
    grid[row][col] = -1
    return helper(grid, row + 1, col) + helper(grid, row - 1, col)  + helper(grid, row, col + 1) + helper(grid, row, col - 1)
    
        
}

func isValid(grid [][]int, row int, col int) bool{
    if row >= len(grid) || row < 0{
        return false
    }
    if col >= len(grid[row]) || col < 0{
        return false
    }
    return true
}
